Sources/OuterCore/Type.swift

```diff
@@ -0,0 +1,246 @@
+/// Type.swift
+///
+/// Copyright 2017, The Silt Language Project.
+///
+/// This project is released under the MIT license, a copy of which is
+/// available in the repository.
+
+import Foundation
+
+public struct NameAndType: Hashable {
+  public let name: String
+  public unowned let type: Type
+
+  public static func ==(lhs: NameAndType, rhs: NameAndType) -> Bool {
+    return lhs.name == rhs.name && lhs.type == rhs.type
+  }
+
+  public var hashValue: Int {
+    return name.hashValue ^ ObjectIdentifier(type).hashValue
+  }
+}
+
+public final class TypeMetadataType: Type {
+  init() {
+    super.init(name: "", type: TypeType.shared)
+  }
+  public override func equals(_ other: Value) -> Bool {
+    return other is TypeMetadataType
+  }
+  public override var hashValue: Int {
+    return 0
+  }
+}
+
+public final class TypeType: Type {
+  static let shared = TypeType()
+
+  init() {
+    let typeType =
+      /// HACK: This only exists to appease the typechecker.
+      unsafeBitCast(nil as TypeType?, to: TypeType.self)
+    super.init(name: "", type: typeType)
+  }
+
+  public override var type: Value {
+    get { return self }
+    set { /* do nothing */ }
+  }
+
+  public override func equals(_ other: Value) -> Bool {
+    return other is TypeType
+  }
+
+  public override var hashValue: Int {
+    return 0
+  }
+}
+
+public final class ArchetypeType: Type {
+  unowned let parent: ParameterizedType
+  let index: Int
+
+  init(parent: ParameterizedType, index: Int) {
+    self.parent = parent
+    self.index = index
+    super.init(name: "", type: TypeType.shared)
+  }
+
+  public override func equals(_ other: Value) -> Bool {
+    guard let other = other as? ArchetypeType else { return false }
+    return parent == other.parent && index == other.index
+  }
+
+  public override var hashValue: Int {
+    return ObjectIdentifier(parent).hashValue ^ index.hashValue ^ 0x374b2947
+  }
+}
+
+public class ParameterizedType: Type {
+  public struct Parameter: Hashable {
+    /*owned */let archetype: ArchetypeType
```

Comment:
This doesn't make sense.  Suppose you've got a function like

```llvm
@foo(%0 : $TypeMetadata, %1 : List[%0], %2 : %TypeMetadata):
```

If the intent is to have a value-parameter pair know about its type metadata, you could have a weak references between parameters linking the pair together.  When we wind up implementing specialization, or if you have a "real monomorphic" value anyways you just have `nil` as your metadata pointer.

Reply:
e.g. `@foo`'s parameter `%1` has a metadata reference to `%2`, and `%0` has a nil metadata reference (or references itself - whatever is cleaner).